README
# Watson IoT – EnOcean Gateway client – Introduction

Gateway to push EnOcean sensor data to Watson IoT platform using Raspberry Pi.

# Dependencies
[Watson IOT Embedded C client](https://github.com/ibm-messaging/iotf-embeddedc)

# Watson IoT – EnOcean Gateway client

### Hardware requirements:
* Raspberry Pi - 2 (model 1 is also expected to work, but recipe not tested on it)
* EnOcean kit containing devices and the EnOcean receiver preferably USB300 which was used for testing this recipe.
### Software requirements:
* Fhem server 5.7 (Last released version: as of 2015-11-15) installed on Raspberry Pi.
* Perl interpreter
  >We will need “perlxs” to build our gateway. To ensure that we can build perlxs modules check for “xsubpp” executable in “bin” directory of your device Perl installation.  
**Note:** In case Perl is not installed on Raspberry Pi (to check type perl -v on Raspberry Pi command line), use the below commands to install Perl  
  * sudo apt-get install perl libdevice-serialport-perl libio-socket-ssl-perl libwww-perl 
  * sudo apt-get install –f 

  >**Note:** It is recommended that you go through the section “Raspberry PI with EnOcean” at http://fhem.de/fhem.html#Links to ensure you have latest information about setting up Fhem with Raspbeery Pi and EnOcean.  
  
* IBM Watson IoT Embedded 'C' client libraries.  
  Link to source code:  https://github.com/ibm-messaging/iotf-embeddedc

# Steps to build Watson IoT – EnOcean Gateway client

### Building the Watson IoT Embedded 'C' client library  
* Download/Clone the repository using the link https://github.com/ibm-messaging/iotf-embeddedc
* Run buildlib.sh supplied to build Watson IoT client shared library “libiotf.so”.

### Building Watson IoT – EnOcean gateway library

* Download/Clone the repository using the link https://github.com/WatsonIoT-Ecosystem/iot_enocean
* Change to the directory where you downloaded the contents in step above. This should be the directory where Makefile.PL resides.
* Run the following commands in sequence  
**Note:** you will need root privileges to do this. On Raspberry pi use “sudo cp ….” to do this copy as root.
   * sudo perl Makefile.PL
     * This will generate the Makefile needed to build the code
   * sudo make
     * This will generate a new shared library named “enoceaniot.so”
   * sudo make install	
     * This command will copy the various files and shared library “enoceaniot.so” to appropriate directories in the Perl installation.  
       <b>Example output:  
       <i>Files found in blib/arch: installing files in blib/lib into architecture dependent library tree  
<i>Installing /usr/local/lib/perl/5.14.2/auto/enoceaniot/enoceaniot.bs  
<i>Installing /usr/local/lib/perl/5.14.2/auto/enoceaniot/enoceaniot.so  
<i>Installing /usr/local/lib/perl/5.14.2/enoceaniot.pm  
<i>Installing /usr/local/lib/perl/5.14.2/auto/enoceaniot/autosplit.ix  
<i>Installing /usr/local/man/man3/enoceaniot.3pm  
<i>Appending installation info to /usr/local/lib/perl/5.14.2/perllocal.pod  

* Copy the Watson IoT Embedded 'C' client library “libiotf.so” to /usr/lib. 
* Copy 98_ENOCEANIOT.pm to /opt/fhem/FHEM directory. This is the directory where you installed Fhem by degault.   

**On completion:** You will have the all the required libraries built and installed at appropriate location.  
To verify run the command *define <some name> ENOCEANIOT* on telnet command line. No error verifies that everything went successfully.




Changes
Revision history for Perl extension enoceaniot.

0.01  Mon Feb  8 06:55:39 2016
	- original version; created by h2xs 1.23 with options
		-O -n enoceaniot ./gtw/gatewayclient.h



Makefile.PL
use 5.014002;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'enoceaniot',
    VERSION_FROM      => 'lib/enoceaniot.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/enoceaniot.pm', # retrieve abstract from module
       AUTHOR         => 'A. U. Thor <pi@>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I. -I../gtw -I../gtw/lib', # e.g., '-I. -I/usr/include/other'
    MYEXTLIB          => '../gtw/build/libiotf.so',
    # Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw(),
              {name=>"CONFIG_FILE_ERROR", macro=>"1"},
              {name=>"MISSING_INPUT_PARAM", macro=>"1"},
              {name=>"QUICKSTART_NOT_SUPPORTED", macro=>"1"});
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'enoceaniot',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


98_ENOCEANIOT.pm
package main;
use strict;
use warnings;
use enoceaniot;

my %ENOCEANIOT_gets = (
	"connect" => "no",
	"publishevent" => ""
);

sub ENOCEANIOT_Initialize($) {
	my ($hash) = @_;

	$hash->{DefFn} = 'ENOCEANIOT_Define';
	$hash->{UndefFn} = 'ENOCEANIOT_Undef';
	$hash->{SetFn} = 'ENOCEANIOT_Set';
	$hash->{GetFn} = 'ENOCEANIOT_Get';
	$hash->{AttrFn} = 'ENOCEANIOT_Attr';
	$hash->{ReadFn} = 'ENOCEANIOT_Read';

	$hash->{AttrList} =
		"orgId gatewayType gatewayId authmethod authtoken";
}

sub ENOCEANIOT_Define($$) {
	my ($hash, $def) = @_;
	my @param = split('[ \t]+', $def);

	#e.g. define myENOIOT ENOCEANIOT
	if(int(@param) < 2) {
		return "too few parameters: define <name> ENOCEANIOT";
	}	

	$hash->{name} = $param[0];

	return undef;
}

sub ENOCEANIOT_Undef($$) {
	my ($hash, $arg) = @_; 
	# nothing to do
	return undef;
}

sub ENOCEANIOT_Set($@) {
	my ($hash, @param) = @_;
	my $rc = 0;
	if(int(@param) < 3)
	{
	   return "set [ENOCEANIOT] needs at least 3 argument. set [ENOCEANIOT] <option> <data>";
	}
	my $name = shift @param;
	my $opt = shift @param;

	if(!defined($ENOCEANIOT_gets{$opt})) 
	{
		my @cList = keys %ENOCEANIOT_gets;
		return "Unknown argument $opt, choose one of " .  join(" ", @cList);
	}
	if($opt eq "connect")
	{
           my $value = join("", @param);
	   if($value eq "yes")
	   {
		if($ENOCEANIOT_gets{$opt} eq "yes")
		{
			return "you are already connected to Watson IoT Platform";
		}
		$rc = initializegateway($attr{$name}{orgId},
				 $attr{$name}{gatewayType},
				 $attr{$name}{gatewayId},
				 $attr{$name}{authmethod},
				 $attr{$name}{authtoken});
		if($rc ne 0)
		{
                   return "Failed to initialize ENOCEANIOT with error $rc";
		}
		
		$rc = connectgateway();
		if($rc ne 0)
		{
                   return "Failed to connect to Watson IoT platform with error $rc";
		}
		$hash->{STATE} = $ENOCEANIOT_gets{$opt} = $value;
		return "Successfully connected to Watson IoT platform";
	   }
	   else
	   {
		#disconnect from Watson IoT Platform
                if($ENOCEANIOT_gets{$opt} eq "yes")
                {
                   $rc = disconnectgateway();
                }
                else
                {
                    return "You are not connected to Watson IoT platform";
                } 
	   }
	}
	if($opt eq "publishevent")
	{
	   if($ENOCEANIOT_gets{"connect"} ne "yes")
           {
              return "You are not connected to Watson IoT platform"; 
           }
           #value is DEVICE TYPE DEVICE ID DATA

           my $DEVTYPE = shift @param;
           my $DEVID = shift @param; 
           my $DATA = join("",@param);	
	   $rc = publishdeviceevent($DEVTYPE,$DEVID,"status","json",$DATA,0);	
	   if($rc ne 0)
	   {
              return "Event publishing failed for $DEVID with error $rc";
	   }
	   $hash->{STATE} = $ENOCEANIOT_gets{$opt} = join($DEVTYPE,$DEVID,$DATA);
	   return "Event succesfully published for $DEVID";
	}

	return "0";
}

sub ENOCEANIOT_Get($@) {
	my ($hash, @param) = @_;
	print @param;	
	if(int(@param) < 1)
	{
		return "get needs at least one argument. get ENOCEANIOT";
	}
	my $name = shift @param;
	#my $opt = shift @param;
	my @cList = keys %ENOCEANIOT_gets;
	return join(" : ", @cList);
}

sub ENOCEANIOT_Attr(@) {
	my ($cmd,$name,$attr_name,$attr_value) = @_;
	if($cmd eq "set") 
	{
		if($attr_name eq "orgId" ||
		   $attr_name eq "gatewayType" ||
		   $attr_name eq "gatewayId" ||
		   $attr_name eq "authmethod" ||
		   $attr_name eq "authtoken")
		{
		   if(length($attr_value) == 0)
		   {
		      my $err = "Invalid value $attr_value for $attr_name.";
		      Log 3, "ENOCEANIOT: ".$err;
		      return $err;
		   }
		} 
		else 
		{
		   return "Unknown attr $attr_name";
		}
	}
	return undef;
}

1;

=pod
=begin html



lib/enoceaniot.pm
package enoceaniot;

use 5.014002;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use enoceaniot ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	CONFIG_FILE_ERROR
	MISSING_INPUT_PARAM
	QUICKSTART_NOT_SUPPORTED
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	CONFIG_FILE_ERROR
	MISSING_INPUT_PARAM
	QUICKSTART_NOT_SUPPORTED
	initializegateway
	connectgateway
	publishgatewayevent
	publishdeviceevent
	disconnectgateway	
);

our $VERSION = '0.01';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&enoceaniot::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('enoceaniot', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

enoceaniot - Perl extension for blah blah blah

=head1 SYNOPSIS

  use enoceaniot;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for enoceaniot, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.

=head2 Exportable constants

  CONFIG_FILE_ERROR
  MISSING_INPUT_PARAM
  QUICKSTART_NOT_SUPPORTED



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

A. U. Thor, E<lt>pi@E<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


=cut
